cmd/snap-exec/main.go

```diff
@@ -0,0 +1,115 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+// for the tests
+var syscallExec = syscall.Exec
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Printf("cannot snap-exec: %s\n", err)
+		os.Exit(1)
+	}
+}
+
+func run() error {
+	var opts struct {
+		Positional struct {
+			SnapApp string `positional-arg-name:"<snapApp>" description:"the application to run, e.g. hello-world.env"`
+		} `positional-args:"yes" required:"yes"`
+
+		Command string `long:"command" description:"use a different command like {stop,post-stop} from the app"`
+	}
+
+	parser := flags.NewParser(&opts, flags.HelpFlag|flags.PassDoubleDash)
+	args, err := parser.Parse()
+	if err != nil {
+		return err
+	}
+
+	snapApp := opts.Positional.SnapApp
+	return snapLaunch(snapApp, opts.Command, args)
+}
+
+func splitSnapApp(snapApp string) (snap, app string) {
+	l := strings.SplitN(snapApp, ".", 2)
+	if len(l) < 2 {
+		return l[0], l[0]
+	}
+	return l[0], l[1]
+}
+
+func findCommand(app *snap.AppInfo, command string) (string, error) {
+	var cmd string
+	switch command {
+	case "stop":
+		cmd = app.StopCommand
+	case "post-stop":
+		cmd = app.PostStopCommand
+	case "":
+		cmd = app.Command
+	default:
+		return "", fmt.Errorf("cannot use %q command", command)
+	}
+
+	if cmd == "" {
+		return "", fmt.Errorf("no %q command found for %q", command, app.Name)
+	}
+	return cmd, nil
+}
+
+func snapLaunch(snapApp, command string, args []string) error {
+	snapName, appName := splitSnapApp(snapApp)
+	info, err := snap.ReadInfo(snapName, &snap.SideInfo{
+		Revision: snap.R(os.Getenv("SNAP_REVISION")),
```

Comment:
Thanks for raising this, I documented this in the code now to make it clearer (hopefully :)